### ISSUE TITLE
[Feature Request] - Only run tests for changed packages in the PR

### ISSUE THREAD

**Issue by petter-eikeland** [flxbl-io/sfp#5]
**Describe the Feature Request**
Only run tests for changed packages in the PR.

I.e. given that validate orgs were created at midnight, and 2 PRs have been merged to master/main changing package A and package B, respectively, both of which containing Apex classes and tests. When creating a 3rd PR, only changing package C (also with Apex classes and tests), `validate` would pick up changes in all 3 packages (A, B and C). In this scenario, `validate` deploys package A, and then executes the corresponding Apex tests. Then the same for package B and then finally package C. Ideally, `validate` would only deploy the (already validated) changes in package A and B (to update the validate org to what's on master/main), and only execute the Apex tests for package C. 
E.g. if `'git diff HEAD origin/master --quiet ' + package["path"] + '/*'` --> then run Apex tests. If not, skip Apex tests for the package.

This additional execution of Apex tests for already validated changes has an especially significant impact on our current cadence given Salesforce's infrastructure issues during release windows.

**Comment by Rocko1204** [comment]
Hi @petter-eikeland The idea would be to implement a validateAgainstBranch command. I've been using it in our pipeline for a year and it describes your feature exactly (i hope :-) ). I could make it available as a branch. Then you could take a look and link it for tests. What does the command do? 

ToDo git job

1. git fetch target branch (main or master)
2. git fetch source branch 
3. checkout source branch 

# Command:

git diff head (source branch) against target branch

so we fetch only changes/packages from the pull request against main branch

## Start validation:

### found unlocked pck change:
- install dependencies from tree 
   - search the deps version on devhub , when found install the deps via package id 
   - when not found search latest pck tree deps version from devhub and install it plus deploy the deps version from path
   - when not found a version deploy from path
- deploy parent pck from path and run testclasses

(with this solution we have no build errors anymore)

### found source pck change:
- deploy parent pck from path and run specific testclasses from pck path

In my case i have split the command for source/diff pck and for unlocked pck because in my case we validate unlocked against a scratch pool and source against a sandbox. And then i run the job in parallel. But its also possible to run it in one command.

**Comment by petter-eikeland** [comment]
Awesome! 

Thank you for the followup 🤩 We would really like to try out this `validateAgainstBranch` mode/command. 

Quick question: Given a validation org is created at midnight, and multiple PRs (changing multiple packages) has been merged. Then, I create a new PR changing one (different) package). Will `validateAgainstBranch` only install the changed dependencies, in addition to the package I changed? Or will it also deploy the changed packages that are not defined as dependencies?

I.e. will my change be validated against all the newest changes on main branch, or only the changes defined by the dependency tree (and my actual change in the PR)?

**Comment by azlam-abdulsalam** [comment]
@Rocko1204  @petter-eikeland  I was very much about to start work on this one, there is a need for a similar command in sfops, especially when used with 'synchronizing' changes from upstream.    I would rather prefer an additional mode added to the existing validate command.

From an implementation perspective, we could use the implementation of  sfp impact topic to understand which packages are impacted ( this only gives the packages impacted on the PR, not necessarily against the tag), and then use that info to run test only for the impacted package, where as others would be deployed
